lightning/src/onion_message/messenger.rs

```diff
@@ -434,17 +443,43 @@ pub enum Destination {
 }
 
 impl Destination {
+	/// Attempts to resolve the [`IntroductionNode::DirectedShortChannelId`] of a
+	/// [`Destination::BlindedPath`] to a [`IntroductionNode::NodeId`], if applicable, using the
+	/// provided [`ReadOnlyNetworkGraph`].
+	pub fn resolve(&mut self, network_graph: &ReadOnlyNetworkGraph) {
```

Comment:
Maybe? I thought about implementing it for `ReadOnlyNetworkGraph`. But it soon became clear that `NodeIdLookUp` is mostly only needed for forwarding, where here the resolution is for creating and sending an onion message (typically via any calls to `OnionMessenger::find_path_and_enqueue_onion_message`) given an unresolved `BlindedPath`.

In `DefaultRouter` we are doing similar resolution to form a hop candidate, only we also check against first hops. So it wasn't clear to me if `NodeIdLookUp` should cover both cases (i.e. forwarding and introduction node) and thus require an additional method.

Note that `create_onion_message` takes a `NodeIdLookUp` for advancing the blinded path by one in case we are the introduction node. However, it doesn't currently do any resolution. Instead, the caller is expected to do that -- either `OnionMessenger` via `DefaultMessageRouter` or anyone else calling `create_onion_message` as a standalone function.

If `NodeIdLookUp` were expanded to include the introduction node resolution, then callers of the standalone `create_onion_message` could pass a `ReadOnlyNetworkGraph` instead of having it fail with `SendError::UnresolvedIntroductionNode`. It would also let `OnionMessenger` resolve unannounced channels of our own, which it seems it currently does not as `DefaultMessageRouter` doesn't have that information. But that would still fail if the peer is disconnected.